src/pmdas/rocestat/pmdarocestat.python

```diff
@@ -0,0 +1,646 @@
+                        self.hw_counter_path.append(path.strip())
+                visit.add(mlx)
+        return active_devices
+
+    # initialize the hardware counters
+    # self.hw_counter_hash['mlnx5_0'][1]['np_cnp_sent'] = <val>
+    # where mlnx5_0 is device, 1 is port, np_cnp_sent is the counter
+    def init_hw_counter(self, active_devices):
+        port_keys = {}
+        port_keys['1'] = 0
+        port_keys['2'] = 0
+        for i in self.hw_counter:
+            token = i.split('/')
+            if (len(token) > 8):
+                mlx = token[4]
+                port = token[6]
+                if mlx not in active_devices:
+                    continue
+                self.hw_counter_path.append(i.strip())
+                counter = token[8].strip()
+                if mlx not in self.hw_counter_hash:
+                    self.hw_counter_hash[mlx] = {}
+                if port not in self.hw_counter_hash[mlx]:
+                    self.hw_counter_hash[mlx][port] = {}
+                self.hw_counter_hash[mlx][port][counter] = 0
+                port_keys[port] += 1
+
+    # get_hw_counters probes the counter periodically and updates the counter stats
+    def get_hw_counters(self, counter_path, hw_counter_vals):
+        for path in counter_path:
+            with open(path, 'r') as fread:
+                ret = fread.read().strip()
+
+            token = path.split('/')
+            mlx, port, counter = token[4], token[6], token[8]
+            hw_counter_vals[mlx][port][counter] = int(ret)
+
+        return hw_counter_vals
+
+    def convert_dict(self, input_dict):
+        output_dict = {}
+        for key1, sub_dict in input_dict.items():
+            if(type(sub_dict) == dict):
+                for key2, value in sub_dict.items():
+                    new_key = f"{key1}_{key2}"
+                    output_dict[new_key] = value
+        return output_dict
+
+    def fetch_mlnx_counters(self):
+        self.hw_counter = self.get_hw_counter_names()
+        active_device = self.get_active_devices()
+
+        self.init_hw_counter(active_device)
+
+        sleep(1)
+
+        self.hw_counter_hash = self.get_hw_counters(self.hw_counter_path, self.hw_counter_hash)
+        return self.convert_dict(self.hw_counter_hash)
+
+
+class ROCESTATPMDA(PMDA):
+    """
+    A ROCE Performance Metrics Domain Agent.
+    Install it and make basic use of it, as follows:
+
+    # $PCP_PMDAS_DIR/rocestat/Install
+    [select python option]
+    $ pminfo -fmdtT rocestat
+    """
+
+    # Cluster definitions
+
+    mlnx_counters = {}
+    nic_stats = {}
+
+    def register_metrics(self, name):
+        """
+        Register Metrics
+        """
+        pm_type_32 = c_api.PM_TYPE_U32
+        pm_type_64 = c_api.PM_TYPE_U64
+        pm_type_double = c_api.PM_TYPE_DOUBLE
+        pm_type_str = c_api.PM_TYPE_STRING
+
+        null_indom = c_api.PM_INDOM_NULL
+        inst_sem = c_api.PM_SEM_INSTANT
+        count_sem = c_api.PM_SEM_COUNTER
+
```

Comment:
There's several unused variables here.  count_sem in particular looks like it should be used for a number of the metrics.